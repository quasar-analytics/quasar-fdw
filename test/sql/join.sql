/* We are comparing JOIN explanations (and verifying outputs)
 * for tables with use_remote_estimate on and off */
/* Big joins are merge joins */
EXPLAIN (COSTS off) SELECT * FROM zips_re z1, zips_re z2 WHERE z1.city = z2.city;
SELECT * FROM zips_re z1, zips_re z2 WHERE z1.city = z2.city LIMIT 10;
/* Hash join with no sort on large table joining smaller one */
EXPLAIN (COSTS off) SELECT * FROM zips_re z1, zips_re z2 WHERE z1.pop > 60000 AND z1.city = z2.city;
SELECT * FROM zips_re z1, zips_re z2 WHERE z1.pop > 60000 AND z1.city = z2.city LIMIT 10;
/* Nested loop join when a tiny number of records */
EXPLAIN (COSTS off) SELECT * FROM zips_re z1, zips_re z2 WHERE z1.pop > 60000 AND z1.state = 'MA' AND z1.city = z2.city;
SELECT * FROM zips_re z1, zips_re z2 WHERE z1.pop > 60000 AND z1.state = 'MA' AND z1.city = z2.city;
/* Cross Join */
EXPLAIN (COSTS off) SELECT * FROM smallzips CROSS JOIN nested;
SELECT * FROM smallzips CROSS JOIN nested LIMIT 2;
/* Outer Join */
EXPLAIN (COSTS off) SELECT * FROM smallzips z1 LEFT OUTER JOIN zips_missing z2 ON z1.city = z2.missing;
SELECT * FROM smallzips z1 LEFT OUTER JOIN zips_missing z2 ON z1.city = z2.missing LIMIT 2;
EXPLAIN (COSTS off) SELECT * FROM smallzips z1 RIGHT OUTER JOIN zips_missing z2 ON z1.city = z2.missing;
SELECT * FROM smallzips z1 RIGHT OUTER JOIN zips_missing z2 ON z1.city = z2.missing LIMIT 2;
/* zips_re state field has a join_rowcount_estimate of 500 so it will use Hash join on some small joins */
EXPLAIN (COSTS off) SELECT * FROM zips_re z1, zips_re z2 WHERE z1.state = z2.state AND z1.city IN ('BARRE', 'AGAWAM');
