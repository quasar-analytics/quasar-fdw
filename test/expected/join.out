/* We are comparing JOIN explanations (and verifying outputs)
 * for tables with use_remote_estimate on and off */
/* Big joins are merge joins */
EXPLAIN (COSTS off) SELECT * FROM zips_re z1, zips_re z2 WHERE z1.city = z2.city;
                              QUERY PLAN                               
-----------------------------------------------------------------------
 Merge Join
   Merge Cond: ((z1.city)::text = (z2.city)::text)
   ->  Sort
         Sort Key: z1.city
         ->  Foreign Scan on zips_re z1
               Quasar query: SELECT "city", "pop", "state" FROM "zips"
   ->  Sort
         Sort Key: z2.city
         ->  Foreign Scan on zips_re z2
               Quasar query: SELECT "city", "pop", "state" FROM "zips"
(10 rows)

SELECT * FROM zips_re z1, zips_re z2 WHERE z1.city = z2.city LIMIT 10;
    city     |  pop  | state | city  |  pop  | state 
-------------+-------+-------+-------+-------+-------
 BARRE       |  4546 | MA    | BARRE |  4546 | MA
 BARRE       |  4546 | MA    | BARRE | 14994 | VT
 AGAWAM      | 15338 | MA    | BARRE |  4546 | MA
 AGAWAM      | 15338 | MA    | BARRE | 14994 | VT
 BLANDFORD   |  1240 | MA    | BARRE |  4546 | MA
 BLANDFORD   |  1240 | MA    | BARRE | 14994 | VT
 BELCHERTOWN | 10579 | MA    | BARRE |  4546 | MA
 BELCHERTOWN | 10579 | MA    | BARRE | 14994 | VT
 CUSHMAN     | 36963 | MA    | BARRE |  4546 | MA
 CUSHMAN     | 36963 | MA    | BARRE | 14994 | VT
(10 rows)

/* Hash join with no sort on large table joining smaller one */
EXPLAIN (COSTS off) SELECT * FROM zips_re z1, zips_re z2 WHERE z1.pop > 60000 AND z1.city = z2.city;
                                          QUERY PLAN                                           
-----------------------------------------------------------------------------------------------
 Hash Join
   Hash Cond: ((z2.city)::text = (z1.city)::text)
   ->  Foreign Scan on zips_re z2
         Quasar query: SELECT "city", "pop", "state" FROM "zips"
   ->  Hash
         ->  Foreign Scan on zips_re z1
               Quasar query: SELECT "city", "pop", "state" FROM "zips" WHERE (("pop" > 60000))
(7 rows)

SELECT * FROM zips_re z1, zips_re z2 WHERE z1.pop > 60000 AND z1.city = z2.city LIMIT 10;
    city     |  pop  | state |    city     |  pop  | state 
-------------+-------+-------+-------------+-------+-------
 FLORENCE    | 66990 | SC    | FLORENCE    | 27939 | MA
 WESTMINSTER | 77965 | CA    | WESTMINSTER |  6191 | MA
 FRAMINGHAM  | 65046 | MA    | FRAMINGHAM  | 65046 | MA
 LEXINGTON   | 69179 | NC    | LEXINGTON   | 28994 | MA
 BRISTOL     | 60670 | CT    | BRISTOL     | 21625 | RI
 BRISTOL     | 60670 | CT    | BRISTOL     |  4288 | NH
 WASHINGTON  | 62924 | DC    | WASHINGTON  |   628 | NH
 ALTON       | 67604 | TX    | ALTON       |  2939 | NH
 BRISTOL     | 60670 | CT    | BRISTOL     |  1220 | ME
 WASHINGTON  | 62924 | DC    | WASHINGTON  |  1261 | ME
(10 rows)

/* Nested loop join when a tiny number of records */
EXPLAIN (COSTS off) SELECT * FROM zips_re z1, zips_re z2 WHERE z1.pop > 60000 AND z1.state = 'MA' AND z1.city = z2.city;
                                                   QUERY PLAN                                                   
----------------------------------------------------------------------------------------------------------------
 Nested Loop
   ->  Foreign Scan on zips_re z1
         Quasar query: SELECT "city", "pop", "state" FROM "zips" WHERE (("pop" > 60000)) AND (("state" = 'MA'))
   ->  Foreign Scan on zips_re z2
         Quasar query: SELECT "city", "pop", "state" FROM "zips" WHERE ((:p1 = "city"))
(5 rows)

SELECT * FROM zips_re z1, zips_re z2 WHERE z1.pop > 60000 AND z1.state = 'MA' AND z1.city = z2.city;
    city    |  pop  | state |    city    |  pop  | state 
------------+-------+-------+------------+-------+-------
 FRAMINGHAM | 65046 | MA    | FRAMINGHAM | 65046 | MA
(1 row)

/* Cross Join */
EXPLAIN (COSTS off) SELECT * FROM smallzips CROSS JOIN nested;
                                                                              QUERY PLAN                                                                              
----------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Nested Loop
   ->  Foreign Scan on smallzips
         Quasar query: SELECT "city", "pop", "state" FROM "smallZips"
   ->  Materialize
         ->  Foreign Scan on nested
               Quasar query: SELECT "topObj"."midObj"."botObj"."a" AS "a", "topObj"."midObj"."botObj"."b" AS "b", "topObj"."midObj"."botObj"."c" AS "c" FROM "nested"
(6 rows)

SELECT * FROM smallzips CROSS JOIN nested LIMIT 2;
    city     |  pop  | state | a | b | c 
-------------+-------+-------+---+---+---
 BELCHERTOWN | 10579 | MA    | m | n | o
 BARRE       |  4546 | MA    | m | n | o
(2 rows)

/* Outer Join */
EXPLAIN (COSTS off) SELECT * FROM smallzips z1 LEFT OUTER JOIN zips_missing z2 ON z1.city = z2.missing;
                                 QUERY PLAN                                 
----------------------------------------------------------------------------
 Merge Right Join
   Merge Cond: ((z2.missing)::text = (z1.city)::text)
   ->  Sort
         Sort Key: z2.missing
         ->  Foreign Scan on zips_missing z2
               Quasar query: SELECT "city", "missing" FROM "smallZips"
   ->  Sort
         Sort Key: z1.city
         ->  Foreign Scan on smallzips z1
               Quasar query: SELECT "city", "pop", "state" FROM "smallZips"
(10 rows)

SELECT * FROM smallzips z1 LEFT OUTER JOIN zips_missing z2 ON z1.city = z2.missing LIMIT 2;
    city     |  pop  | state | city | missing 
-------------+-------+-------+------+---------
 BELCHERTOWN | 10579 | MA    |      | 
 BARRE       |  4546 | MA    |      | 
(2 rows)

EXPLAIN (COSTS off) SELECT * FROM smallzips z1 RIGHT OUTER JOIN zips_missing z2 ON z1.city = z2.missing;
                                 QUERY PLAN                                 
----------------------------------------------------------------------------
 Merge Left Join
   Merge Cond: ((z2.missing)::text = (z1.city)::text)
   ->  Sort
         Sort Key: z2.missing
         ->  Foreign Scan on zips_missing z2
               Quasar query: SELECT "city", "missing" FROM "smallZips"
   ->  Sort
         Sort Key: z1.city
         ->  Foreign Scan on smallzips z1
               Quasar query: SELECT "city", "pop", "state" FROM "smallZips"
(10 rows)

SELECT * FROM smallzips z1 RIGHT OUTER JOIN zips_missing z2 ON z1.city = z2.missing LIMIT 2;
 city | pop | state |    city     | missing 
------+-----+-------+-------------+---------
      |     |       | BELCHERTOWN | 
      |     |       | BARRE       | 
(2 rows)

